### PR TITLE
Add test (and fixes) for ResourceGroup tags update

### LINKED ISSUE #1452: Updating a resource group fails
Hi, 

I see that this project is marked experimental, but I thought I'd have a look anyway!

I can successfully create a resource group using a definition like this:
```
apiVersion: microsoft.resources.infra.azure.com/v20191001
kind: ResourceGroup
metadata:
  name: test-resource-group
spec:
  location: uksouth
```

But if I try to make a change to it - for instance, to add tags - I see errors like this:
```
  Warning  ReconcileError            2s                 ResourceGroupCtrl  failed to patch with: failed to apply state to Azure with apply failed with: uri: https://management.azure.com/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/deployments/k8s_1606417240_ad2f3275-3019-11eb-9f2b-d6aecc665e74?api-version=2019-10-01, status: 400, body: {"error":{"code":"InvalidTemplate","message":"Deployment template validation failed: 'The template resource 'Microsoft.Resources/resourceGroups/test-resource-group' at line '1' and column '339' has invalid value of 'id' property. Expected value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/resourceGroups/test-resource-group' and actual value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/resourceGroups/test-resource-group'. Please see https://aka.ms/arm-template/#resources for usage details.'.","additionalInfo":[{"type":"TemplateViolation","info":{"lineNumber":1,"linePosition":339,"path":"Properties.template.resources[0]"}}]}}
  Warning  ReconcileError            1s                 ResourceGroupCtrl  failed to patch with: failed to apply state to Azure with apply failed with: uri: https://management.azure.com/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/deployments/k8s_1606417241_adcda1de-3019-11eb-9f2b-d6aecc665e74?api-version=2019-10-01, status: 400, body: {"error":{"code":"InvalidTemplate","message":"Deployment template validation failed: 'The template resource 'Microsoft.Resources/resourceGroups/test-resource-group' at line '1' and column '339' has invalid value of 'id' property. Expected value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/resourceGroups/test-resource-group' and actual value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/resourceGroups/test-resource-group'. Please see https://aka.ms/arm-template/#resources for usage details.'.","additionalInfo":[{"type":"TemplateViolation","info":{"lineNumber":1,"linePosition":339,"path":"Properties.template.resources[0]"}}]}}
  Normal   ResourceHasChanged        0s (x3 over 2s)    ResourceGroupCtrl  resource in state "Succeeded" has changed and spec will be applied to Azure
  Warning  ReconcileError            0s                 ResourceGroupCtrl  failed to patch with: failed to apply state to Azure with apply failed with: uri: https://management.azure.com/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/deployments/k8s_1606417242_ae6bfe3a-3019-11eb-9f2b-d6aecc665e74?api-version=2019-10-01, status: 400, body: {"error":{"code":"InvalidTemplate","message":"Deployment template validation failed: 'The template resource 'Microsoft.Resources/resourceGroups/test-resource-group' at line '1' and column '339' has invalid value of 'id' property. Expected value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/providers/Microsoft.Resources/resourceGroups/test-resource-group' and actual value '/subscriptions/852785fa-eac5-46b0-b864-1c395db07668/resourceGroups/test-resource-group'. Please see https://aka.ms/arm-template/#resources for usage details.'.","additionalInfo":[{"type":"TemplateViolation","info":{"lineNumber":1,"linePosition":339,"path":"Properties.template.resources[0]"}}]}}
```

Is this currently unsupported function, or am I doing something wrong?

Thanks!

### PR DESCRIPTION
**Pull request by Porges**
Closes #1452, sorta.

**What this PR does / why we need it**:

Adds a U to the CRUD tests for resource groups, which updates a resource group to add tags to it.

As part of this, I found an issue with the way conflicts (409) are handled in `CreateDeployment`; we should back off and wait for the deployment to exit the conflicting state (currently assumed to be deleting). This would cause test failures as the deletion of the initial “create” deployment would still be pending when the test went to do the “update” deployment.

There are other issues remaining in the controller that I have marked with two new TODOs:
- during monitor deployment we state that the resource provisioning state is equivalent to the deployment state which is confusing and a bit wrong; when the deployment is deleting then the resource provisioning state is not deleting
- we need to think about how to handle other types of conflict in `CreateDeployment`

Also added is a `timeRemaining` helper that can be used with `gomega.Eventually` to give better behaviour on timeouts.

**Special notes for your reviewer**:

Thanks @matthchr for helping diagnose!

**How does this PR make you feel**:
gif

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains tests

### CODE REVIEW

File: hack/generated/controllers/helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	"testing"
+	"time"
+)
+
+// remainingTime returns how long is left until test timeout,
+// and can be used with gomega.Eventually to get better failure behaviour
+//
+// (If you hit the deadline 'go test' aborts everything and dumps
+// the current task stacks to output. If gomega.Eventually hits its
+// timeout it will produce a nicer error message and stack trace.)
```
Comment (theunrepentantgeek) [inline hack/generated/controllers/helpers_test.go:18]: Nice!

File: hack/generated/pkg/reconcilers/azure_deployment_reconciler.go [hunk 2]
```diff
@@ -379,21 +381,19 @@ func (r *AzureDeploymentReconciler) StartDeleteOfResource(ctx context.Context) (
...
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "creating empty status for %q", resource.GetId())
+	}
 
-		retryAfter, deleteErr = r.ARMClient.BeginDeleteResource(ctx, resource.GetId(), resource.Spec().GetApiVersion(), emptyStatus)
-		if deleteErr != nil {
-			return errors.Wrapf(deleteErr, "deleting resource %q", resource.Spec().GetType())
-		}
+	// retryAfter = ARM can tell us how long to wait for a DELETE
+	retryAfter, err := r.ARMClient.BeginDeleteResource(ctx, resource.GetId(), resource.Spec().GetApiVersion(), emptyStatus)
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/reconcilers/azure_deployment_reconciler.go:390]: We don't seem to ever use `retryAfter`

**Review by theunrepentantgeek — COMMENTED** [review]
One minor comment - if that's not an issue, ping me and I'll approve.

File: hack/generated/pkg/reconcilers/azure_deployment_reconciler.go (hunk 2 above)
Comment (Porges) [reply hack/generated/pkg/reconcilers/azure_deployment_reconciler.go:390]: it is used in the return value down below (if it wasn't used at all then the linter would complain!)

Comment (theunrepentantgeek) [reply hack/generated/pkg/reconcilers/azure_deployment_reconciler.go:390]: I searched (using the browswer) and didn't find it. I believe you though. 😁

**Review by theunrepentantgeek — APPROVED** [review]

**Pull request merged**